gensim/similarities/docsim.py

```diff
@@ -186,11 +186,27 @@ def __init__(self, output_prefix, corpus, num_features, num_best=None, chunksize
         `self.num_best = 10` before doing a query.
 
         """
+        self.is_shard = False  # To indicate if shard directory has been created, if not then shard and index files are stored in temp
         if output_prefix is None:
             # undocumented feature: set output_prefix=None to create the server in temp
             self.output_prefix = utils.randfname(prefix='simserver')
         else:
-            self.output_prefix = output_prefix
+            try:
+                self.output_prefix = os.path.join(os.path.dirname(output_prefix), 'shard', '')   # '' to get the trailing slash
+                if not os.path.exists(self.output_prefix):
+                    os.makedirs(self.output_prefix)
+                    logger.info("created %s directory . Both shard and index files are stored at this location. Please use \
+                    this entire folder for portability across different systems. ", self.output_prefix)
+                    self.is_shard = True  # helps while saving the index file as we don't want to create directories in temp
+                else:
+                    logger.warning("It seems there is already a file named shard in location %s . Please ensure  \
+                    that there is no other files named shard in the provided location. Starting similarity index under temp folder", os.path.join(os.path.dirname(output_prefix), ''))
+                    self.output_prefix = utils.randfname(prefix='simserver')
+            except:
+                logger.warning("Failed to create shard folder in %s. Please ensure that you have required permission to create  \
```

Comment:
I do not understand the PR description, the phrasing and formatting is hard to parse.

Either way, if user says "store under directory X", we cannot store under directory Y silently. That's just wrong.

If we wish to only support storing to directory Y (and nowhere else), then we shouldn't give the user even the option of saying "store to directory X". The API should be clear.

The problem is not failing if the user asks for some impossible action -- we should fail, explicitly and clearly. The problem is if the API is confusing or does unintuitive clandestine stuff.